### PR TITLE
fix(medusa): Optimize Cart totals calculation

### PR DESCRIPTION
**Pull request by srindom**
**What**

The existing totals calculations are extremely heavy and perform an enormous amount of duplicate work. The changes here remove large parts of the overhead and improves response times for cart endpoints up to 30x.

### CODE REVIEW

File: integration-tests/api/__tests__/taxes/manual-taxes.js [hunk 1]
```diff
@@ -25,7 +25,7 @@ describe("Manual Cart Taxes", () => {
   beforeAll(async () => {
     const cwd = path.resolve(path.join(__dirname, "..", ".."))
     dbConnection = await initDb({ cwd })
-    medusaProcess = await setupServer({ cwd })
+    medusaProcess = await setupServer({ cwd, verbose: true })
```
Comment (olivermrbl) [inline integration-tests/api/__tests__/taxes/manual-taxes.js:28]: todo: Flip flag
```suggestion
    medusaProcess = await setupServer({ cwd, verbose: false })
```

File: integration-tests/plugins/.yarnrc.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+npmRegistryServer: "http://localhost:4873"
```
Comment (olivermrbl) [inline integration-tests/plugins/.yarnrc.yml:1]: question: Do we need this or is it specific to your local env?

File: packages/medusa/src/services/cart.ts [hunk 3]
```diff
@@ -327,6 +343,33 @@ class CartService extends TransactionBaseService {
...
+    return await this.decorateTotals_(raw, totalsToSelect, totalsConfig)
+  }
+
+  private async retrieveNew(
+    cartId: string,
+    options: FindConfig<Cart> = {}
+  ): Promise<Cart> {
+    const manager = this.manager_
+    const cartRepo = manager.getCustomRepository(this.cartRepository_)
+    const validatedId = validateId(cartId)
```
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:365]: suggestion(non-blocking): Remove as it has no effect

File: packages/medusa/src/services/totals.ts [hunk 4]
```diff
@@ -983,13 +949,21 @@ class TotalsService extends TransactionBaseService {
    * @param cartOrOrder - the cart or order to get gift card amount for
    * @return the gift card amount applied to the cart or order
    */
-  async getGiftCardTotal(cartOrOrder: Cart | Order): Promise<{
+  async getGiftCardTotal(
+    cartOrOrder: Cart | Order,
+    opts: { gift_cardable?: number } = {}
```
Comment (olivermrbl) [inline packages/medusa/src/services/totals.ts:954]: suggestion(non-blocking): Rename to be more descriptive

**Review by olivermrbl — COMMENTED** [review]
so nice!

File: integration-tests/plugins/.yarnrc.yml (hunk 2 above)
Comment (srindom) [reply integration-tests/plugins/.yarnrc.yml:1]: It is required to run `medusa-dev` in the `integration-tests/plugins` dir locally; not specific to my env only. 

It is similar to `integration-tests/api/.yarnrc.yml`.

**Review by olivermrbl — APPROVED** [review]

**Pull request merged**